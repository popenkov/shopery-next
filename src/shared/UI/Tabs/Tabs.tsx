'use client';
import { useEffect, useState } from 'react';

import cn from 'classnames';

import cls from './Tabs.module.scss';
import { Text } from '..';

interface SliderProps<T> {
  data: T[];
}

export interface TabItem<T> {
  id: string;
  label: string;
  content: T[];
}

interface TabsProps<T> {
  tabs: TabItem<T>[];
  onTabChange?: (value: string) => void;
  tabContent: (item: T) => JSX.Element;
  sliderContent?: (props: SliderProps<T>) => JSX.Element;
}

export const Tabs = <T extends unknown>({
  tabs,
  onTabChange,
  tabContent,
  sliderContent,
}: TabsProps<T>) => {
  const [selectedTab, setSelectedTab] = useState<string>(tabs?.[0].id);
  const [tabContentState, setTabContentState] = useState<T[]>([]);

  const handleTabChange = (value: string) => {
    setSelectedTab(value);

    onTabChange?.(value);
  };

  useEffect(() => {
    const activeContent = tabs?.find((t) => t.id === selectedTab)?.content;
    setTabContentState(activeContent || []);
  }, [selectedTab, tabs]);

  return (
    <div className={cn(cls.tabs, 'section__content')}>
      <div className={cn(cls.labels, 'container')}>
        <div className={cls.labelsContent}>
          <div className={cls.scrollContainer}>
            {tabs?.map((tab) => {
              const isActive = selectedTab === tab.id;
              return (
                <button
                  className={cn(cls.label, {
                    [cls.active]: isActive,
                  })}
                  onClick={() => handleTabChange(tab.id)}
                  key={tab.id}
                >
                  {tab.label}
                </button>
              );
            })}
          </div>
        </div>
      </div>
      <div className={cls.panes}>
        <div className={cls.pane}>
          <div className="grid grid__four-items  d-desktop container">
            {tabContentState?.map((item) => {
              return tabContent(item);
            })}
          </div>
          {sliderContent && (
            <div className="container mobile-slider d-mobile">
              {sliderContent({ data: tabContentState })}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};
