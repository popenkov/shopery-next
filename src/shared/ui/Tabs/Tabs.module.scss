@import '@/shared/styles/mixins/text-styles';
@import '@/shared/styles/mixins/flexbox';
@import '@/shared/styles/breakpoints';

.tabs {
  max-width: 100vw;
  overflow: hidden;
}

.labels {
  @media screen and (max-width: $screen-ti) {
    position: relative;
    width: 100%;

    &::before {
      position: absolute;
      top: 0;
      left: 0;
      z-index: 2;
      display: block;
      width: 16px;
      height: 100%;
      background: linear-gradient(
        to left,
        rgb(4 119 210 / 0%),
        var(--green-gray-50)
      );
      content: '';
    }

    &::after {
      position: absolute;
      top: 0;
      right: 0;
      z-index: 2;
      display: block;
      width: 16px;
      height: 100%;
      background: linear-gradient(
        to right,
        rgb(4 119 210 / 0%),
        var(--green-gray-50)
      );
      content: '';
    }
  }
}

.labelsContent {
  @media screen and (max-width: $screen-ti) {
    position: relative;
    width: 100%;
    height: 100%;
    padding: 0 20px;
    overflow: auto hidden;
    scrollbar-width: none;
    -webkit-overflow-scrolling: touch;
  }
}

.scrollContainer {
  @include flex-center;

  position: relative;
  gap: 12px;
  outline: none;

  @media screen and (max-width: $screen-ti) {
    width: max-content;
  }
}

.label {
  @include body-m(500);

  position: relative;
  flex-shrink: 0;
  padding: 0;
  padding: 8px 12px 6px;
  color: var(--gray-500);
  background-color: inherit;
  border-top-width: 0;
  border-right-width: 0;
  border-bottom: 2px solid transparent;
  border-left-width: 0;
  cursor: pointer;
  transition: border-color var(--transition-long), color var(--transition-long);

  &:hover {
    color: var(--primary);
  }

  &.active {
    color: var(--primary);
    border-color: var(--primary);
    cursor: default;
  }

  &:not(:last-child) {
    margin-right: 15px;

    &::after {
      position: absolute;
      top: 50%;
      right: -15px;
      width: 1px;
      height: 16px;
      background-color: var(--green-gray-200);
      transform: translateY(-50%);
      content: '';
    }
  }
}

.panes {
  position: relative;
  border-top: 0;
}

.labels + .panes {
  margin-top: 47px;
}

.pane {
  width: 100%;
  max-width: 100vw;
}
