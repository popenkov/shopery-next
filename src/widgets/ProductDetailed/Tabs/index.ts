export { ProductDetailedTabs } from './ui/Tabs';
