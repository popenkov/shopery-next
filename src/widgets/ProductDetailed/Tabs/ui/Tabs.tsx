import { FC } from 'react';
import cn from 'classnames';

import { VideoAd } from 'entities/Advertisement/ui/VideoAd/VideoAd';
import { TProductDetailed } from 'entities/Product';
import { ProductReview } from 'entities/Review';
import { Video } from 'entities/Video';
import Tabs from 'shared/ui/Tabs/Tabs';

import cls from './Tabs.module.scss';

type Props = {
  descriptions: string;
  ad: TProductDetailed['ad'];
  specs: TProductDetailed['specs'];
  feedback: TProductDetailed['feedback'];
};

export const ProductDetailedTabs: FC<Props> = ({ ad, descriptions, specs, feedback }) => {
  const TABS = [
    {
      label: 'Descriptions',
      content: (
        <div className={cls.tabsContent}>
          <div className={cls.tabsDescription} dangerouslySetInnerHTML={{ __html: descriptions }} />
          <VideoAd
            video={<Video data={ad.video} hasOverlay={false} buttonType="secondary" />}
            adData={ad.advantages}
            className={cls.tabsAside}
          />
        </div>
      ),
    },
    {
      label: 'Additional Information',
      content: (
        <div className={cls.tabsContent}>
          <div className={cls.tabsDescription}>
            <div className="product-page__specs">
              {specs.map((spec, index) => {
                return (
                  <div className="product-page__spec" key={`spec/${index}`}>
                    <span className="product-page__spec-title">{spec.title}</span>
                    <span className="product-page__spec-value">
                      {spec.value.map((value, index) => {
                        return (
                          <span
                            className="product-page__spec-value-item"
                            key={`spec-value/${index}`}
                          >
                            {value.title}
                          </span>
                        );
                      })}
                    </span>
                  </div>
                );
              })}
            </div>
          </div>
          <VideoAd
            video={<Video data={ad.video} hasOverlay={false} buttonType="secondary" />}
            adData={ad.advantages}
            className={cls.tabsAside}
          />
        </div>
      ),
    },
    {
      label: 'Customer Feedback',
      content: (
        <div className={cn(cls.tabsContent, cls.tabReviews)}>
          {feedback.map((item, index) => {
            return <ProductReview data={item} key={`feedback/${index}`} />;
          })}
        </div>
      ),
    },
  ];

  return <Tabs tabs={TABS} />;
};
