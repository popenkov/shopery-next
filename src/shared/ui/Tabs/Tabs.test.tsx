import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Tabs } from './Tabs';

describe('Tabs component', () => {
  it('renders correctly', () => {
    const tabs = [
      { label: 'Tab 1', content: <div>Content 1</div> },
      { label: 'Tab 2', content: <div>Content 2</div> },
    ];
    const { container } = render(<Tabs tabs={tabs} />);
    expect(container).toMatchSnapshot();
  });

  it('renders correct number of tabs', () => {
    const tabs = [
      { label: 'Tab 1', content: <div>Content 1</div> },
      { label: 'Tab 2', content: <div>Content 2</div> },
      { label: 'Tab 3', content: <div>Content 3</div> },
    ];
    const { getAllByRole } = render(<Tabs tabs={tabs} />);
    const tabButtons = getAllByRole('button');
    expect(tabButtons.length).toBe(3);
  });

  it('renders correct content for selected tab', () => {
    const tabs = [
      { label: 'Tab 1', content: <div>Content 1</div> },
      { label: 'Tab 2', content: <div>Content 2</div> },
    ];
    const { getByText } = render(<Tabs tabs={tabs} />);
    expect(getByText('Content 1')).toBeInTheDocument();
  });

  it('switches to correct tab on click', () => {
    const tabs = [
      { label: 'Tab 1', content: <div>Content 1</div> },
      { label: 'Tab 2', content: <div>Content 2</div> },
    ];
    const { getAllByRole, getByText } = render(<Tabs tabs={tabs} />);
    const tabButtons = getAllByRole('button');
    fireEvent.click(tabButtons[1]);
    expect(getByText('Content 2')).toBeInTheDocument();
  });

  it('has correct class names', () => {
    const tabs = [
      { label: 'Tab 1', content: <div>Content 1</div> },
      { label: 'Tab 2', content: <div>Content 2</div> },
    ];
    const { container } = render(<Tabs tabs={tabs} />);
    expect(container.querySelector('.tabs')).toBeInTheDocument();
    expect(container.querySelector('.labels')).toBeInTheDocument();
    expect(container.querySelector('.labelsContent')).toBeInTheDocument();
    expect(container.querySelector('.scrollContainer')).toBeInTheDocument();
    expect(container.querySelector('.label')).toBeInTheDocument();
    expect(container.querySelector('.active')).toBeInTheDocument();
    expect(container.querySelector('.panes')).toBeInTheDocument();
    expect(container.querySelector('.pane')).toBeInTheDocument();
    expect(container.querySelector('.paneContent')).toBeInTheDocument();
  });
});
