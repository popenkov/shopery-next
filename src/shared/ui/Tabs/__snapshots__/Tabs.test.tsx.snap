// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component renders correctly 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="labels"
    >
      <div
        class="labelsContent"
      >
        <div
          class="scrollContainer"
        >
          <button
            class="label active"
          >
            Tab 1
          </button>
          <button
            class="label"
          >
            Tab 2
          </button>
        </div>
      </div>
    </div>
    <div
      class="panes"
    >
      <div
        class="pane"
      >
        <div
          class="paneContent"
        >
          <div>
            Content 1
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
